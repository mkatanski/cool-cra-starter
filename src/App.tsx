import React from "react";
import { MainStyle } from "styles/global";
import { ThemeProvider } from "components/theme/ThemeProvider";
import { ThemedContainer } from "components/theme/ThemedContainer";
import { MainContainerTheme as LightTheme } from "styles/themes/light";
import { MainContainerTheme as DarkTheme } from "styles/themes/dark";
import { TopBar } from "components/TopBar";
import { LinkButton } from "components/LinkButton";

function App() {
  return (
    <ThemeProvider>
      <ThemedContainer themes={{ light: LightTheme, dark: DarkTheme }}>
        <MainStyle />
        <TopBar>
          <TopBar.Container>
            <LinkButton>Home</LinkButton>
          </TopBar.Container>
          <TopBar.Container>
            <LinkButton>Page 1</LinkButton>
          </TopBar.Container>
          <TopBar.Container>
            <LinkButton>Page 2</LinkButton>
          </TopBar.Container>
          <TopBar.Container alignRight>
            <LinkButton>Switch Theme</LinkButton>
          </TopBar.Container>
        </TopBar>
        <div className="App">
          <LinkButton>Login view</LinkButton>
        </div>
      </ThemedContainer>
    </ThemeProvider>
  );
}

export default App;
