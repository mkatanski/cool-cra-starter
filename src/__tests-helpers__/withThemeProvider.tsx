import React from "react";
import { ThemeContext } from "components/theme/ThemeProvider";
import { StyledTheme } from "styles/_types";
import { ThemeName } from "config/themes";
import { ThemeProvider } from "styled-components";
import { palette } from "styles/palette";

const TestTheme: StyledTheme = {
  container: {
    background: "container.background",
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },

  button: {
    text: {
      default: "palette.holly",
      hover: "palette.outerSpace",
    },
    background: {
      default: "palette.azalea",
      hover: "palette.charm",
    },
  },

  input: {
    background: {
      default: "palette.azalea",
      hover: "palette.azalea",
    },
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },

  table: {
    background: "palette.beautyBush",
    border: "palette.wewak",
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },

  tableHeader: {
    background: "palette.beautyBush",
    border: "palette.wewak",
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },

  tableRowEven: {
    background: "palette.wewak",
    border: "palette.wewak",
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },

  tableRowOdd: {
    background: "palette.beautyBush",
    border: "palette.wewak",
    text: {
      primary: "palette.holly",
      inverse: "palette.beautyBush",
      secondary: "palette.outerSpace",
    },
  },
};

export const withThemeProvider = (Component: React.ReactNode) => (
  <ThemeContext.Provider
    value={{
      themeMode: ThemeName.Light,
      changeTheme: jest.fn(),
    }}
  >
    <ThemeProvider
      theme={{
        elements: TestTheme,
        palette: palette,
      }}
    >
      {Component}
    </ThemeProvider>
  </ThemeContext.Provider>
);
