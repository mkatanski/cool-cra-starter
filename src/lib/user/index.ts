export * from "./userContexts";
export * from "./userProvider";
export * from "./userReducer";
export * from "./userActions";
export * from "./userState";
