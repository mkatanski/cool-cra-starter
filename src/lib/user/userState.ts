export type UserState = {
  isLoggedIn: boolean;
  username: string;
  accessToken?: string;
  refreshToken?: string;
};

export const defaultState: UserState = {
  isLoggedIn: true,
  username: "",
};
