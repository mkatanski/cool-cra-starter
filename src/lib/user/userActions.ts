import { UserState } from "./userState";

export type SetUserAction = {
  type: "SET_USER";
  user: UserState;
};

export type UpdateUserAction = {
  type: "UPDATE_USER";
  user: Partial<UserState>;
};

export type UserAction = SetUserAction | UpdateUserAction;

export const clearUserData = (): SetUserAction => ({
  type: "SET_USER",
  user: {
    isLoggedIn: false,
    username: "",
  },
});

export const setUserData = (username: string): SetUserAction => ({
  type: "SET_USER",
  user: {
    isLoggedIn: true,
    username,
  },
});
