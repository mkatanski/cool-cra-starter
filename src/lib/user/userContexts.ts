import { UserState, defaultState } from "./userState";
import { UserAction } from "./userActions";
import { makeStoreContexts } from "utils/store";

const [StateContext, DispatchContext] = makeStoreContexts<
  UserState,
  UserAction
>(defaultState);

export const UserStateContext = StateContext;
export const UserDispatchContext = DispatchContext;
