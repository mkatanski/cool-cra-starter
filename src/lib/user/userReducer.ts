import { UserAction } from "./userActions";
import { UserState } from "./userState";

export const userReducer = (
  state: UserState,
  action: UserAction
): UserState => {
  switch (action.type) {
    case "SET_USER":
      return action.user;
    case "UPDATE_USER":
      return {
        ...state,
        ...action.user,
      };
  }
};
