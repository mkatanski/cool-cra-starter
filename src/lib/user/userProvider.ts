import { UserStateContext, UserDispatchContext } from "./userContexts";
import { userReducer } from "./userReducer";
import { makeStoreProvider } from "utils/store";
import { defaultState } from "./userState";

export const UserProvider = makeStoreProvider({
  StateContext: UserStateContext,
  DispatchContext: UserDispatchContext,
  reducer: userReducer,
  defaultState,
});
