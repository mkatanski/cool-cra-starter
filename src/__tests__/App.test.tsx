/* eslint-disable react/display-name */
import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";
import App from "App";

import { UserStateContext } from "lib/user";
import Routes from "../routes";

import { withThemeProvider } from "__tests-helpers__/withThemeProvider";

jest.mock("routes", () => jest.fn());

jest.mock("react-router-dom", () => ({
  BrowserRouter: (props: React.PropsWithChildren<{}>) => (
    <div id="BrowserRouter">{props.children}</div>
  ),
}));

jest.mock("components/fallbacks/LoadingScreen", () => ({
  LoadingScreen: jest.fn(() => {
    return <div id="LoadingScreen" />;
  }),
}));

describe("App", () => {
  it("should render App component", async () => {
    (Routes as jest.Mock).mockImplementationOnce(() => <div>Routes</div>);

    const tree = TestRenderer.create(
      <UserStateContext.Provider value={{ isLoggedIn: true, username: "test" }}>
        <App />
      </UserStateContext.Provider>
    );
    expect(tree).toMatchSnapshot();
  });

  it("should render a loading screen as a fallback", async () => {
    (Routes as jest.Mock).mockImplementationOnce(() => {
      throw new Promise((r) => {
        r();
      });
    });
    const tree = TestRenderer.create(withThemeProvider(<App />));
    expect(tree).toMatchSnapshot();
  });
});
