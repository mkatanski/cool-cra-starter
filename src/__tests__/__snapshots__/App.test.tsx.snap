// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render App component 1`] = `
<div
  className="App"
>
  <div
    id="BrowserRouter"
  >
    <div>
      Routes
    </div>
  </div>
</div>
`;

exports[`App should render a loading screen as a fallback 1`] = `
<div
  className="App"
>
  <div
    id="BrowserRouter"
  >
    <div
      id="LoadingScreen"
    />
  </div>
</div>
`;
